### DISCUSSION TITLE
Technical discussion: Replacing the "Place order" button with a custom component

### DISCUSSION THREAD

**Discussion by opr** [woocommerce/woocommerce#45064] [WooCommerce, Blocks, and Site Editing]
# Introduction
This discussion is to collect ideas around an extensibility API to replace Checkout Block's "Place order" button with a custom button, or other component. The aim of the discussion is to identify any technical issues around doing this, as well as highlight concerns and issues that may arise when implementing this API.

A related discussion is here: [Customizting the "Place order" button on the Block Checkout](https://github.com/woocommerce/woocommerce-blocks/discussions/9337) where further reasoning is provided to show this API is something needed by third party plugins.

# Why is this needed?

For some context on how the checkout block works here is an article covering that: [How The Checkout Block Processes An Order](https://developer.woocommerce.com/2022/10/06/how-the-checkout-block-processes-an-order/)

To summarise, the existing button kicks off the checkout process and runs all of our internal checks (such as form validation) and submits the checkout request to the API. Plugins can hook into various events during this process such as the validation step before processing begins.

## Processing outside the Checkout block
This API may be required when a payment gateway needs to do some off-site processing to decide what specific button to show users, based on some off-site processing. There is also the case where the user pressing the button is taken off-site to complete some type of processing, e.g. logging into an account with a provider, but an express method is not appropriate, for example if the store captures some details through a custom field, or is using extensions such as EU VAT Number which can't be completed when using an express payment method. 

## Custom Branding
There is currently no API to programatically "brand" the buttons, and this is not possible outside of a CSS approach by the developer. This may be desirable for some plugins if they have different brands across different territories. Admittedly, we could enable some APIs to achieve this branding, however if the branding to apply is calculated in a "black-box" by the third party's servers, using logic they don't necessarily want to expose to the client then maybe the only approach would be allowing a custom component in place of our checkout button.

## Other reasons?
If folks have any additional use-cases where this API would be useful, I invite you to join this discussion and present your ideas and thoughts.

# Pros & Cons

## Pros
- Allows full freedom to customise and show specific elements in the Checkout block.
- Removes the need for us to implement more focused APIs to achieve specific button style changes, reducing maintenance.
- More complex processing can take place in the third-party's custom code that may not be achievable through blocks, similar gain to the above, i.e. fewer APIs to maintain.

## Cons
- Allows full freedom to customise and show specific elements in the Checkout block, developers could create buttons that depart from our desired look and feel/UX.
- We will need to very carefully manage the conditions under which this custom component can be rendered to ensure no single plugin can spoil the Checkout process for other payment methods by removing the button, or otherwise breaking the checkout process.

# Requirements

## It cannot be misused by other competing payment gateways to spoil the experience of shoppers when using a different gateway.

To implement this, I propose the [`PaymentMethodConfig` interface](https://github.com/woocommerce/woocommerce-blocks/blob/b38c6e91546b36a34e517668129bbc2794d74f7f/assets/js/types/type-defs/payments.ts#L117C4-L130) is expanded to include a new option. To prevent misuse and to encourage developers to think a little longer about what they are doing, I propose we call the new property: `dangerouslyReplacePlaceOrderButton` or something similar. 

Because this custom button would be registered against a specific payment method, we would engineer it so that it can only show when that payment method is selected by the shopper. When another payment method is selected, we would unmount the custom component.

## It must allow the extension to selectively render its custom button based on some conditions

For example, only show if all items in the cart are shippable, or if the total is over £100.

To achieve this, we must pass the cart data to the payment method to allow it to decide whether to render the button or not.

It may be useful to ask extensions to supply _another_ property on the [`PaymentMethodConfig`](https://github.com/woocommerce/woocommerce-blocks/blob/b38c6e91546b36a34e517668129bbc2794d74f7f/assets/js/types/type-defs/payments.ts#L117C4-L130) called `shouldDangerouslyReplacePlaceOrderButton`. This would be a function whose result would determine whether the component supplied in `dangerouslyReplacePlaceOrderButton` should be displayed. Similar to `canMakePayment`.

This function may be useful because, if `dangerouslyReplacePlaceOrderButton` is a React component, the logic to determine whether to show the custom button or not must take place within that component, and we must pass the cart data via props to that component. We won't know until after that component has been rendered whether or not we should display it or use the Checkout Block's "Place order" button instead.

If we use a plain function that returns a boolean value, that allows extensions to explicitly indicate that they want to show the custom button, reducing the potential for it to accidentally show if we base it solely on the output of a React component. I feel the tradeoff in "inconvenience" for the developer but increased safety for the product and the shopper is OK. Here's an example.

```tsx
const shouldDangerouslyReplacePlaceOrderButton = ( cartData: Cart ) => {
  return parseInt( cartData.totals.total_price, 10 ) > 100;
};

const dangerouslyReplacePlaceOrderButton = <CustomOrderButtonComponent />;

registerPaymentMethod( {
  ...otherConfigValues,
  dangerouslyReplacePlaceOrderButton,
  shouldDangerouslyReplacePlaceOrderButton,
} );
```

## We must fall back to the Checkout Block's "Place order button" if the custom button doesn't render or if it errors

We also need to ensure that we fall back to the Checkout Block's button in case the custom component does not render anything, or if it errors. This is to ensure the shopper has the ability to continue with their order.

We need to explore ideas about what to do if the custom button is critical to the payment method's functionality, and it would not work without the custom button. Would we have to switch the user to a different payment method? How can we solve this issue?

# Summary & Expectations

To summarise, I am proposing a new extensibility point, added to `registerPaymentMethod` which allows extensions to display their own custom component in place of the Checkout Block's "Place order" button. We would only allow this to be shown if the payment method this button was registered with is the active one.

My expectations on this discussion are to discuss the following:
- Besides what is listed above, can you think of any other use-cases for this API?
- Does this sound like a sensible approach?
- Can you think of any alternative solutions that are less extreme than fully replacing the button?
- Are the "safety measures" I mentioned good enough? What else can we implement to ensure the shopper is able to complete their order if something goes wrong with the custom button?

**Comment by opr** [comment]
We also suggested a possible approach using the existing `PaymentMethodConfig` interface, however it looks like [this doesn't solve every use-case](https://github.com/woocommerce/woocommerce-blocks/discussions/9337#discussioncomment-5852575). I will mention it here anyway so we have a record of it.

1. User selects "Custom payment gateway"
2. Your plugin styles the Checkout Block's "Place order button" with the correct branding for their location via some API we implement
3. Your payment method registers an observer for the `onPaymentSetup` (previously [`onPaymentProcessing`](https://github.com/woocommerce/woocommerce-blocks/blob/4eae7caf0815972bbf1ff4ef7bee6c6dc688d011/docs/internal-developers/block-client-apis/checkout/checkout-api.md#payment-method-events-context)) event. This observer contains two things:

  i. Code to open a modal which contains all the logic you need to get the user authenticated/ready to pay.
   
  ii. A `Promise` that does not resolve until some condition is met, thereby suspending execution of the checkout flow.

4. 
```js
const onPaymentSetupUnsubscribe = onPaymentSetup( async () => {
	let isLoginComplete = false;
	const setLoginComplete = () => {
		isLoginComplete = true;
	};
	window.customNamespace.setLoginComplete = setLoginComplete;
	await new Promise( ( resolve ) => {
		window.open(
			'/popup.html',
			'_blank',
			'width=600,height=600,popup=yes'
		);
		setInterval( () => {
			if ( isLoginComplete ) {
				resolve();
			}
		}, 1000 );
	} );
} );
```

```html
Welcome to the popup login page
<script>
    const onLoginClick = () => {
        window.opener.setLoginComplete();
        window.close();
    }
</script>
<button onClick="onLoginClick()">Log in</button>
```
5. When the button in the modal is clicked, it triggers the callback on the `opener` window, which changes the `isLoginComplete` flag to true, causing the promise to resolve.

**Comment by nerrad** [comment]
> It may be useful to ask extensions to supply another property on the [PaymentMethodConfig](https://github.com/woocommerce/woocommerce-blocks/blob/b38c6e91546b36a34e517668129bbc2794d74f7f/assets/js/types/type-defs/payments.ts#L117C4-L130) called shouldDangerouslyReplacePlaceOrderButton. This would be a function whose result would determine whether the component supplied in dangerouslyReplacePlaceOrderButton should be displayed. Similar to canMakePayment.

I'm wary that this overcomplicates things, and have concerns about the naming. 

First, I don't understand the need for having a separate boolean for whether the component displays or not. IMO that should be determined by two criteria:

1. The button override component is present for that gateway.
2. The gateway is active (`canMakePayment` returns true) and selected by the shopper.

I'd also suggest implementing the logic for displaying the button in [this component](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/components/cart-checkout/place-order-button/index.tsx) where you could also pass in relevant props for the payment method's custom button to consume and signal with.

Second, if this is an official API that we're accommodating, I don't think we need to make it scary. There are other ways (some you already noted) that the API can use to introduce safety through constraints.

**Reply by opr** [reply]
@nerrad thanks for your comments.

> First, I don't understand the need for having a separate boolean for whether the component displays or not. IMO that should be determined by two criteria:

> The button override component is present for that gateway.
> The gateway is active (canMakePayment returns true) and selected by the shopper.

My goal with this was to ensure that the extension had to be very explicit about replacing the button. After thinking about this more, I don't think this is truly necessary. 

The thought process was, if the extension is not very explicit about displaying it, then we would show the existing "Place order" button, however the existing button probably wouldn't work correctly if the payment gateway required an iframe, so maybe there's not much point having this fallback.

> I'd also suggest implementing the logic for displaying the button in [this component](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/base/components/cart-checkout/place-order-button/index.tsx) where you could also pass in relevant props for the payment method's custom button to consume and signal with.

That seems like the perfect place to do this.

> Second, if this is an official API that we're accommodating, I don't think we need to make it scary. There are other ways (some you already noted) that the API can use to introduce safety through constraints.

Thanks for your feedback here, after considering it again, I agree. If we're implementing and supporting it we can introduce other  measures to ensure the checkout can fall back to a safe state if something goes wrong.

**Comment by mikejolley** [comment]
I added a comment here to that iframe use case: https://github.com/woocommerce/woocommerce-blocks/discussions/9337#discussioncomment-5903650

I may be misunderstanding things but it seems dangerous to allow a gateway to take payment via its own button because it would not be certain that blocks has updated/processed the order from the client to the server. We do sync data as the order progresses but I wonder if thats enough to trigger proper validation etc. We don't want a scenario where payment is completed and then the order is rejected by validation rules.

In the legacy checkout we implemented a "pay page" step for some gateways like this where after placing the order (in pending state) one could then complete payment. I don't recall which gateways did it but it may have been inserted on the order received page after placing the order.

I wonder if gateways with the limitation of iframes should be able to adopt a similar approach—a separate page after placing the order where payment is facilitated which then updates the order from 'pending' to 'processing'.

**Reply by opr** [reply]
> I may be misunderstanding things but it seems dangerous to allow a gateway to take payment via its own button because it would not be certain that blocks has updated/processed the order from the client to the server. We do sync data as the order progresses but I wonder if thats enough to trigger proper validation etc. We don't want a scenario where payment is completed and then the order is rejected by validation rules.

I don't think you're misunderstanding, I think this a real danger, and our syncs during form completion aren't enough to fully validate an order, there could be further validation checks that only fire when the place order button is pressed, for example extensions using the `onPaymentProcessing` observer, or even server-side using the various hooks that happen during checkout.

> In the legacy checkout we implemented a "pay page" step for some gateways like this where after placing the order (in pending state) one could then complete payment. I don't recall which gateways did it but it may have been inserted on the order received page after placing the order.
>
> I wonder if gateways with the limitation of iframes should be able to adopt a similar approach—a separate page after placing the order where payment is facilitated which then updates the order from 'pending' to 'processing'.

I think what you're suggesting is a good feature, but maybe not a direct solution for this problem. The point of allowing extensions to _replace_ the button is because the button branding, text etc. may change based on several criteria that we are unaware of, for example geolocation. Because it may change per client, this is not something that can just be overridden with static CSS, there needs to be some way of dynamically setting this.

One option is to determine the branding to show on the server, during page load. When you know what button to show, send this data sent to the client via an `IntegrationInterface` which a client-side script would interpret and apply the correct styling to the button.

**Reply by InpsydeNiklas** [reply]
> I added a comment here to that iframe use case: https://github.com/woocommerce/woocommerce-blocks/discussions/9337#discussioncomment-5903650

@mikejolley I messed up [my response](https://github.com/woocommerce/woocommerce-blocks/discussions/9337#discussioncomment-5906224) to your comment by posting it as another comment. Sorry, first time GitHub discussions 😅


> We don't want a scenario where payment is completed and then the order is rejected by validation rules.

Valid concern. But while the payment provider initializes the payment process before the WooCommerce order is validated and created, the payment can usually be captured at a later stage.
In our case, this means that capturing the funds from an approved payment state only happens after the WooCommerce validation is passed and the order is created. 
Should anything go wrong during the WooCommerce validation, normally, no funds would be debited.
The custom iframe payment button may also trigger a server-side validation of the checkout form fields before starting the payment process. So it's not easy to end up in a scenario where funds are debited, but the WC order failed to validate/create. Of course, there could be exceptions with certain payment providers.

> In the legacy checkout we implemented a "pay page" step for some gateways like this where after placing the order (in pending state) one could then complete payment. I don't recall which gateways did it but it may have been inserted on the order received page after placing the order.

Performing the payment from the order-received page would be possible, but it would negatively impact the user experience. As previously mentioned, the button may adjust to every visitor and serve as an advertisement tool to get the buyer to use this method. Showing this button only after the order confirmation would limit the potential to impact the buyer's choice of payment method before they settled. The custom payment button may actually also consist of multiple buttons, depending on varying factors.
So this would be a downgrade in user experience compared to what is possible now on the shortcode Checkout page (replacing the "Place order" button).

> One option is to determine the branding to show on the server, during page load. When you know what button to show, send this data sent to the client via an IntegrationInterface which a client-side script would interpret and apply the correct styling to the button.

@opr Custom styling may be useful for certain applications, but it doesn't directly apply to the scenario with the custom iframe buttons. The server is not aware of the button styling either since the button exists in an iframe on another server. Rebuilding the button manually would be a nightmare to maintain, so there is no other viable alternative to embedding this iframe button on the page, ideally instead of the "Place order" button.

**Comment by ralucaStan** [comment]
Here is a ticket with a similar request. https://github.com/woocommerce/woocommerce/issues/43839

**Comment by 0xBeycan** [comment]
@opr hey is it possible now?

**Reply by opr** [reply]
No but we're working on it!
